### DISCUSSION TITLE
Extended PWM Limits 700-2300us

### DISCUSSION THREAD

**Discussion by lesharodin** [ExpressLRS/ExpressLRS#2522] [Ideas]
There is a working modification by [@AirNovel ](https://github.com/AirNovel), that give us a wider pwm output for more servo angle (180 degree on SG90).
He achieved it by adding one line, that makes all channels expect 3rd (that usually used for throttle) to output 700-2300us PWM
` if (ch != 2) us = fmap(us, 988, 2012, 700, 2300); `
https://github.com/AirNovel/ExpressLRS_wide_PWM/blob/d84898feb56fd6646cd7336bfe110663aa29cedb/src/lib/ServoOutput/devServoOutput.cpp#L78

Would be cool if someone woud make smth like an option in web/lua to set this to a specific channel. 
PWM recievers are getting very popular now, and people coming to ELRS from frsky/spectrum/etc is confused why there is no E.Limits support
https://edgetx.gitbook.io/edgetx-user-manual/edgetx-user-manual/user-manual-for-color-screen-radios/model-settings/inputs-mixes-and-outputs/outputs

**Comment by Andron1967** [comment]
![123](https://github.com/ExpressLRS/ExpressLRS/assets/52330659/5444e0f1-17fa-445f-b496-39e28760ff1f)
![1234](https://github.com/ExpressLRS/ExpressLRS/assets/52330659/8d632253-4f6d-499e-a1aa-fab84f4cba5a)
are we talking about these receivers?

**Reply by lesharodin** [reply]
@Andron1967 Any PWM reciever

**Comment by Andron1967** [comment]
Did I understand correctly that this modification changes the values of all channels except channel 3 (gas)?

**Comment by r1000ru** [comment]
Hello. Maybe not engage in hard code, but implement full-fledged mixing?

I played a little as a programmer and implemented mixing with the following functionality:
1. Each PWM output can be mixed with each channel
2. For each output/channel mix, the scaling and offset are indicated.
3. Each PWM output can be assigned a minimum and maximum value.

Below is an example of a rather complex mixing, playground available onhttps://onlinegdb.com/E4mgk_iw2 :
```c
    /**
     * Sample of configuration
     * Airplane aka AtomRC Swordfish:
     * - Flapperons (Aelerons with flaps function)
     * - V-tail
     * - Two engine with different thrust
     * - Camera panoramic servo
     * Channels:
     * - 0 - Aelerons
     * - 1 - Elevator
     * - 2 - Throttle
     * - 3 - Rudder
     * - 4 - Arm
     * - 5 - Flaps
     * - 6 - Camera panoramic
     * PWM-outputs:
     * - 0 - engine left
     * - 1 - engine right
     * - 2 - flapperon left
     * - 3 - flapperon right
     * - 4 - v-tail left
     * - 5 - v-tail right
     * - 6 - panoramic servo (270 deg)
     */
    
    // Left engine value mix throttle, arm, rudder
    pwmSettings[0].mixers[2].scale = 100;       // 100% of thrust (-500...500) + 
    pwmSettings[0].mixers[3].scale = 20;        // 20% of rudder (-100...100) +
    pwmSettings[0].mixers[4].scale = 110;       // 110% of arm (-550...+550) +
    pwmSettings[0].mixers[4].offset = -550;     // -550 arm offset (-1100...0)
    
    // Right engine value mix throttle, arm, rudder
    pwmSettings[1].mixers[2].scale = 100;       // 100% of thrust (-500...500) + 
    pwmSettings[1].mixers[3].scale = -20;       // 20% of rudder (+100...-100) +
    pwmSettings[1].mixers[4].scale = 110;       // 110% of arm (-550...+550) +
    pwmSettings[1].mixers[4].offset = -550;     // -550 arm offset (-1100...0)
    
    // Left flapperon value mix aeleron, flaps
    pwmSettings[2].mixers[0].scale = 100;       // 100% of aeleron (-500...+500) + 
    pwmSettings[2].mixers[5].scale = 20;        // 20% of flaps (-100...+100) + 
    pwmSettings[2].mixers[5].offset = 100;      // +100 offset of flaps(0...+200)

    // Right flapperon value mix aeleron, flaps
    pwmSettings[3].mixers[0].scale = 100;       // 100% of aeleron (-500...+500) + 
    pwmSettings[3].mixers[5].scale = -20;       // -20% of flaps (+100...-100) + 
    pwmSettings[3].mixers[5].offset = -100;     // -100 offset of flaps(0...-200)

    // Left V-tail value mix elevator, rudder
    pwmSettings[4].mixers[1].scale = 50;        // 50% of elevator (-250...+250) + 
    pwmSettings[4].mixers[3].scale = 50;        // 50% of rudder (-250...+250) 

    // Right V-tail value mix elevator, rudder
    pwmSettings[5].mixers[1].scale = -50;       // -50% of elevator (+250...-250) + 
    pwmSettings[5].mixers[3].scale = 50;        // 50% of rudder (-250...+250) 

    // Panoramic camera
    pwmSettings[6].mixers[6].scale = 200;       // 200% of panoramic (-1000...+1000)
    pwmSettings[6].limit.min = 500;
    pwmSettings[6].limit.max = 2500;

```

**Comment by Ncerzzk** [comment]
This function would be very useful when the handset could not do mixing configuration.
While this situation is not common, I'm not sure elrs community  would accept this or not.

**Comment by Andron1967** [comment]
As an airplane pilot, I would like this function to be used in the firmware

**Comment by Ncerzzk** [comment]
This extending function and mixer has been implemented in this branch:
https://github.com/Ncerzzk/ExpressLRS/tree/add_mixer

example could be found here:
https://github.com/ExpressLRS/ExpressLRS/pull/2671

**Comment by ss8688** [comment]
I have a 500us-2500us servo

**Comment by Ncerzzk** [comment]
mixer implement update: add configure gui in index.html rather than hardware.
![PixPin_2024-06-15_17-53-48](https://github.com/ExpressLRS/ExpressLRS/assets/9284611/99c0dc92-84e5-4207-847c-63f42bcb1deb)

will be used in a simple gyro mixer(on going)

**Comment by atomGit** [comment]
i also would like to see a 500-2500us range (RM ER8 here) - many of the [programmable servos from AGFRC](https://www.agf-rc.com/Standard-Servo-c101788/) and other companies use this range and AGFRC sells white-label and custom servos to many other companies - typically they are used in ground, air and water vehicles

**Reply by findus4fun** [reply]
Same here on a sailboat using agrfc b26clm. I would like to be able to use 500 us to 2500 us to achievr 180 degrees. I could reducr the length of the servo arm to have more power.

**Reply by tschaerni** [reply]
Extended Limits (or rather programmable via web GUI) would also be highly appreciated in the Combat Robotics community. Especially with the smaller weight classes (150g, 450g) you are quite space and weight limited, so adding a "servo signal extender" is often not really an option.

**Comment by yarda** [comment]
SG90 is very popular servo. With the stable firmware it's going only to +-45 degrees. At the moment it isn't blocker for us, because we use it on boat, but it would be great to have the full +-90 degrees range configurable without modding the firmware.
